pkg/local_object_storage/blobstor/fstree/head.go

```diff
@@ -0,0 +1,218 @@
+
+const (
+	_ = iota
+	fieldObjectID
+	fieldObjectSignature
+	fieldObjectHeader
+	fieldObjectPayload
+
+	smallObjectSizeLimit = 100 << 10 // 100 KB
+)
+
+// Head returns an object's header from the storage by address without reading the full payload.
+func (t *FSTree) Head(addr oid.Address) (*objectSDK.Object, error) {
+	p := t.treePath(addr)
+
+	f, err := os.Open(p)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return nil, logicerr.Wrap(apistatus.ErrObjectNotFound)
+		}
+		return nil, fmt.Errorf("read file %q: %w", p, err)
+	}
+	defer f.Close()
+
+	obj, err := t.extractHeaderOnly(addr.Object(), f)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return nil, logicerr.Wrap(apistatus.ErrObjectNotFound)
+		}
+		return nil, fmt.Errorf("extract object header from %q: %w", p, err)
+	}
+
+	return obj, nil
+}
+
+// extractHeaderOnly reads the header of an object from a file.
+func (t *FSTree) extractHeaderOnly(id oid.ID, f *os.File) (*objectSDK.Object, error) {
+	var comBuf [combinedDataOff]byte
+	n, err := io.ReadFull(f, comBuf[:])
+	if err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			var obj objectSDK.Object
+			return &obj, obj.Unmarshal(comBuf[:n])
+		}
+		return nil, err
+	}
+
+	thisOID, l := parseCombinedPrefix(comBuf)
+	if thisOID == nil {
+		st, err := f.Stat()
+		if err != nil {
+			return nil, err
+		}
+		sz := st.Size()
+		if sz > math.MaxInt {
+			return nil, errors.New("too large file")
+		}
+		if sz < smallObjectSizeLimit {
+			return t.headSmallObject(f, comBuf[:n], sz)
+		}
+		return t.readUntilPayload(f, comBuf[:n])
+	}
+
+	for {
+		if bytes.Equal(thisOID, id[:]) {
+			if l < smallObjectSizeLimit {
+				return t.headSmallObject(f, nil, int64(l))
+			}
+			return t.readUntilPayload(f, nil)
+		}
+		if _, err := f.Seek(int64(l), io.SeekCurrent); err != nil {
+			return nil, err
+		}
+
+		_, err := io.ReadFull(f, comBuf[:])
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+				return nil, logicerr.Wrap(apistatus.ErrObjectNotFound)
+			}
+			return nil, err
+		}
+
+		thisOID, l = parseCombinedPrefix(comBuf)
+		if thisOID == nil {
+			return nil, errors.New("malformed combined file")
+		}
+	}
+}
+
+// headSmallObject reads a small object from the file, decompresses it if necessary,
+// and returns the object with its payload cut off.
+func (t *FSTree) headSmallObject(f *os.File, initial []byte, size int64) (*objectSDK.Object, error) {
+	data := make([]byte, size)
+	copy(data, initial)
+	_, err := io.ReadFull(f, data[len(initial):])
+	if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return nil, fmt.Errorf("read: %w", err)
+	}
+
+	data, err = t.Decompress(data)
+	if err != nil {
+		return nil, fmt.Errorf("decompress file data: %w", err)
+	}
+
+	var obj objectSDK.Object
+	err = obj.Unmarshal(data)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal object: %w", err)
+	}
+	return obj.CutPayload(), nil
```

Comment:
So 100K has the same performance wrt this? Even on spinning disks? I'd be a bit more conservative with ~32-64K here. Then some code can be shared between `Get()` and `Head()` (read/decompress/unmarshal is all the same).

Notice that if you're to have 16K in a buffer already it'll all be easier in most cases.